### PR TITLE
feat(Slider)!: support form

### PR DESCRIPTION
**Pull request by ValeraS**

### CODE REVIEW

File: src/components/Slider/BaseSlider/BaseSlider.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ import './BaseSlider.scss';
 const b = block('base-slider');
 
 type BaseSliderProps = {stateModifiers: StateModifiers} & Omit<
-    SliderProps,
+    SliderProps<number | [number, number]>,
```
Comment (Arucard89) [inline src/components/Slider/BaseSlider/BaseSlider.tsx:16]: Why is this done here? this leads to the need to specify the @ts-expect-error directive at line 25

File: src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -2,14 +2,15 @@
 
 import React from 'react';
 
-import debounce from 'lodash/debounce';
-
+import {useControlledState} from '../../hooks';
```
Comment (Arucard89) [inline src/components/Slider/Slider.tsx:5]: is it necessary here? BaseSlider component has out-of-the-box props for controlled and uncontrolled behaviour without this hook

File: src/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -103,41 +108,59 @@ export const Slider = React.forwardRef(function Slider(
...
+                            <React.Fragment>
+                                {handle}
+                                <input
+                                    ref={inputRef}
+                                    type="hidden"
+                                    name={name}
+                                    form={form}
+                                    value={handleProps.value}
+                                    disabled={disabled}
+                                />
```
Comment (Arucard89) [inline src/components/Slider/Slider.tsx:133]: is it correct to put hidden input inside the handler? 
may be it will be better to move it outside the component and connect ti Slider thru Slider's props (onChange, value, etc)?

File: src/components/Slider/types.ts [hunk 4]
```diff
@@ -8,7 +8,7 @@ export type SliderValue = number | [number, number];
 
 export type RcSliderValueType = number | number[];
 
-export type SliderProps<ValueType = number | [number, number]> = {
+export interface SliderProps<ValueType = number | [number, number]> extends DOMProps, QAProps {
```
Comment (Arucard89) [inline src/components/Slider/types.ts:11]: why did you change type to interface?

**Comment by Arucard89** [comment]
@ValeraS hi. 

I see deep refactoring of basic component which adds breaking change and very specific case of usage .

May be you consider to make HOC (withForm) or new FormSlider component based on Slider component? in this case, you could add changes you need and move away from deep refactoring of the Slider and avoid breaking changes.

@korvin89 , @amje , what do you think about it?

**Comment by korvin89** [comment]
@Arucard89 
> I see deep refactoring of basic component which adds breaking change

I think it is okay for the major version I suppose

> May be you consider to make HOC (withForm) or new FormSlider wrapper component based on Slider component? 

We are trying to add this functionality to all uikit controls into the core code

**Comment by Arucard89** [comment]
> @Arucard89
> 
> > I see deep refactoring of basic component which adds breaking change
> 
> I think it is okay for the major version
> 
> > May be you consider to make HOC (withForm) or new FormSlider wrapper component based on Slider component?
> 
> We are trying to add this functionality to all uikit controls into the core code

oh, I see. There was no description, so I didn't know that it is planned work

File: src/components/Slider/Slider.tsx (hunk 3 above)
Comment (Arucard89) [reply src/components/Slider/Slider.tsx:133]: > is it correct to put hidden input inside the handler? may be it will be better to move it outside the component and connect ti Slider thru Slider's state?

Also it is necessary to take into account the probability of two values appearing [number, number]. At my point it will be easier to manage such case.

For example like this:

```ts
...
const inputRef = useFormResetHandler({
    initialValue: Array.isArray(innerValue) ? innerValue[0] : innerValue,
    onReset: handleReset,
});
const inputSecondRef = useFormResetHandler({
    initialValue: Array.isArray(innerValue) ? innerValue[1] : innerValue,
    onReset: handleReset,
});
...
return (
        <div
            {...filterDOMProps(otherProps)}
            id={id}
            className={b(null, className)}
            ref={ref}
            style={style}
            data-qa={qa}
        >
            <input
                ref={inputRef}
                ... //
            />
            {Array.isArray(innerValue)? <input 
                ref={inputSecondRef}
                ... 
                /> : null}
            <div className={b('top', {size, hasTooltip})}></div>
            <BaseSlider
        ... />
        ...

```

File: src/components/Slider/Slider.tsx [hunk 5]
```diff
@@ -30,46 +31,26 @@ export const Slider = React.forwardRef(function Slider(
...
         tabIndex,
         className,
+        style,
         qa,
         apiRef,
         'aria-label': ariaLabelForHandle,
         'aria-labelledby': ariaLabelledByForHandle,
+        name,
+        form,
+        id,
```
Comment (Arucard89) [inline src/components/Slider/Slider.tsx:48]: Are these props(`name`, `id`) fit to case with range slider, when value is [number, number]?

File: src/components/Slider/Slider.tsx (hunk 3 above)
Comment (ValeraS) [reply src/components/Slider/Slider.tsx:133]: Normally, a slider component uses input with a type range under the hood, but not in this library. IMHO, integration with forms should be a base functionality for the slider component.

File: src/components/Slider/Slider.tsx (hunk 2 above)
Comment (ValeraS) [reply src/components/Slider/Slider.tsx:5]: Yes, because we control state of the BaseSlider component.

File: src/components/Slider/Slider.tsx (hunk 5 above)
Comment (ValeraS) [reply src/components/Slider/Slider.tsx:48]: Yes. `id` goes to container. And several inputs with the same name are valid (see tests).

File: src/components/Slider/Slider.tsx (hunk 3 above)
Comment (Arucard89) [reply src/components/Slider/Slider.tsx:133]: @ValeraS , yes I agreed with you. And I suggest to put input near the BaseSlider instead of inside handle rendering function. I think it will be ore logical. 
Input still will be inside the Slider component. Because (in my opinion) input looks strange inside render prop.

File: src/components/Slider/BaseSlider/BaseSlider.tsx (hunk 1 above)
Comment (Arucard89) [reply src/components/Slider/BaseSlider/BaseSlider.tsx]: Sorry for misunderstanding) 
I'd like to know why do we need this changes in types at all?)

File: src/components/Slider/Slider.tsx (hunk 3 above)
Comment (ValeraS) [reply src/components/Slider/Slider.tsx:133]: There is should be an input for each handle, so it's convenient to render the input alongside the handle.

Comment (Arucard89) [reply src/components/Slider/Slider.tsx:133]: ok, thanks)

File: src/components/Slider/BaseSlider/BaseSlider.tsx (hunk 1 above)
Comment (ValeraS) [reply src/components/Slider/BaseSlider/BaseSlider.tsx]: Changes give more correct types for `value` and `onUpdate`. Before this, type of value always `number | number[]` for `BaseSlider` and `number | [number | number]` for `Slider`. After this, if `value` has type 'number' then `onUpdate` also receives only values of type `number`. The same thing applies to the `[number, number]` type.

Comment (Arucard89) [reply src/components/Slider/BaseSlider/BaseSlider.tsx]: ok, I see. thanks)

**Review by amje — APPROVED** [review]

File: src/components/Slider/Slider.tsx [hunk 6]
```diff
@@ -95,47 +69,71 @@ export const Slider = React.forwardRef(function Slider(
...
+    const inputRef = useFormResetHandler({initialValue: innerValue, onReset: handleReset});
+
+    const stateModifiers: StateModifiers = {
+        size,
+        error: validationState === 'invalid' && !disabled,
+        disabled,
+        'tooltip-display': innerState.tooltipDisplay,
+        rtl: direction === 'rtl',
+        'no-marks': Array.isArray(marks) ? marks.length === 0 : marks === 0,
+    };
```
Comment (Arucard89) [inline src/components/Slider/Slider.tsx:95]: why did you remove memoization? 
it can help prevent creating object on each render and could prevent rerenders inside components

File: src/components/Slider/Slider.tsx [hunk 7]
```diff
@@ -95,47 +69,71 @@ export const Slider = React.forwardRef(function Slider(
...
+
+    const handleRender = (
```
Comment (Arucard89) [inline src/components/Slider/Slider.tsx:97]: why did you remove memoization? 
It could prevent some rerenders

File: src/components/Slider/__tetsts__/Slider.form.test.tsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                </form>
+            );
+        }
+
+        render(<Test />);
+        const form = screen.getByTestId('form');
+        expect(form).toHaveFormValues({slider: ['5', '10']});
+
+        const sliderHandle = screen.getAllByRole('slider')[1];
+        fireEvent.keyDown(sliderHandle, {key: 'ArrowRight', keyCode: 39, code: 'ArrowRight'});
```
Comment (ogonkov) [inline src/components/Slider/__tetsts__/Slider.form.test.tsx:114]: May be use userEvent.keyboard? Could be shorter

Comment (ValeraS) [reply src/components/Slider/__tetsts__/Slider.form.test.tsx:114]: `userEvent.keyboard` does not set `keyCode` for the event, but `rs-slider` checks it.

**Comment by amje** [comment]
@ValeraS @Arucard89 Going to merge this PR as we needed this feature. Please continute your discussion about memoization later if you want to.
